datasets.py

```diff
@@ -31,47 +38,188 @@ def create_mnist_datasets(
+    for feature_name in feature_names:
+        num_channels = _get_spoken_digit_feature_channels(feature_name)
+        signature[feature_name] = tf.TensorSpec(shape=(None, num_bins, num_channels), dtype=tf.float32)
+        padded_shapes[feature_name] = (None, num_bins, num_channels)
+        
+    signature = [signature]
+        
+    if load_labels:
+        signature.append(tf.TensorSpec(shape=(), dtype=tf.int8))
+        padded_shapes = (padded_shapes, ())
+    
+    return tuple(signature), padded_shapes
+
+
+def _extract_spoken_digit_features_npy(example, config, load_labels):
+    audio = example.audio # this should be a numpy array
+    label = example.label if load_labels else None
+    # TODO data augmentation
+
+    windowed_frame_config = config['dsp']['windowed_frames']
+    stft_features_config = config['dsp']['stft_features']
+    
+    windowed_frame_data = dsp.windowed_frame_analysis(audio, windowed_frame_config)
+    features = dsp.stft_analysis(windowed_frame_data['frames'], stft_features_config)
+    
+    # convert numpy arrays to tensors
+    for feature_name in features:
+        num_channels = _get_spoken_digit_feature_channels(feature_name)
+        feature_frames = features[feature_name].view(f'({num_channels},)float')
+        features[feature_name] = tf.convert_to_tensor(feature_frames, dtype=tf.float32)
+    
+    if label is not None:
+        label = tf.convert_to_tensor(label, dtype=tf.int8)
+        return (features, label)
+    
+    return features
+
+
+def _count_spoken_digit_dataset_npy(data_dir):
+    return sum([1 for _ in data_dir.glob('*.npy')])
+
+
+def _load_spoken_digit_dataset_npy(data_dir, config, load_labels):
+    for data_file in data_dir.glob('*.npy'):
+        data = np.load(data_file)
+        yield _extract_spoken_digit_features_npy(np.rec.array(data, dtype=data.dtype), config, load_labels)
+
+
+def load_spoken_digit_dataset(data_dir, config, load_labels):
+    count = _count_spoken_digit_dataset_npy(data_dir)
+    generator = lambda: _load_spoken_digit_dataset_npy(data_dir, config, load_labels)
+    out_signature, padded_shapes = _get_spoken_digit_signature_and_padded_shape(config, load_labels)
+    all_data = tf.data.Dataset.from_generator(generator, output_signature=out_signature)
+    all_data = all_data.apply(tf.data.experimental.assert_cardinality(count))
+    return all_data, padded_shapes
+
+
+def create_spoken_digit_datasets(
+        data_dir,
+        config,
+        batch_size=32,
+        train_test_split=0.8,
+        load_labels=False,
+        repeat=False,
+) -> Tuple[tf.data.Dataset, tf.data.Dataset]:    
+    all_data, padded_shapes = load_spoken_digit_dataset(data_dir, config, load_labels)
+    total_count = len(all_data)
+
+    train_count = int(train_test_split * total_count)
+    test_count = total_count - train_count
+    
+    train_data = all_data.take(train_count).cache().shuffle(train_count, reshuffle_each_iteration=True)\
+    
+    if repeat == False:
```

Comment:
There are no realistic cases where we'd not want to repeat the dataset when we're creating from a generator. Also the assert below, if ever run, would take a very long time because it essentially would have to iterate over the entire object. I'd remove this, but if you want to keep it I'd set the default `repeat=True`. Also below it seems you've used `repeat` as an int which makes the typing too flexible. Just don't set the `count` in `repeat`, the model trainer will handle things.

Reply:
it's useful for debugging purposes to have a mode where the generator has a finite size, so I do want to keep the option there. 

About the `assert`s, they were helpful while I was writing the code and might help avoid some regressions but if you think it will really slow things down I don't mind removing them now that things are working as expected